src/clj_puppetdb/http.clj

```diff
@@ -24,47 +24,86 @@
   `puppetlabs.ssl-utils.core/pems->ssl-context` function."
   [:ssl-cert :ssl-key :ssl-ca-cert])
 
+(defn- make-client-common
+  [^String host opts]
+  (let [opts (assoc opts :as :stream)
+        info {:host host}
+        info (if-let [ssl-context (:ssl-context opts)]
+               (assoc info :ssl-context ssl-context)
+               info)]
+    (fn
+      ; This arity-overloaded variant of the function is the one called by the `GET` function.
+      ; The `this` parameter refers under normal circumstances to this very function and it is
+      ; needed to be able to call another arity-overloaded variant of the function from the
+      ; function body. When the `this` parameter refers to a different function then it
+      ; effectively overrides the other arity-overloaded variants of the function.
+      ([this path params]
+       (let [query (-> (StringBuilder. host)
+                       (.append path))]
+         (if-not (empty? params)
+           (-> query
+               (.append \?)
+               (.append (-> params
+                            q/params->json
+                            map->query))))
+         ; now call the single argument variant of the function to execute the query
+         (this (.toString query))))
```

Comment:
We're in philosophical territory here but my preference is towards the more "Clojurey" style that Karel alluded to. I think succinctness and simplicity are some of the advantages of Clojure and we risk losing them when transliterating Java. Certainly that'll sometimes be necessary but I'd reserve it for the occasions we know we measurable problems - like the slow reports endpoint thing we've seen this week for example. Doing things the "Clojure way" should also let us take advantage of improved tooling and compilers for free over the coming months/years.

Like I say, philosophical territory more than anything.